In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')

df = pd.read_csv('data.csv', sep=';')

# Baseline Model

In [9]:
# ─────────────────────────────────────────
# BASELINE MODEL 1: Logistic Regression
# ─────────────────────────────────────────

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score,
    f1_score, classification_report, confusion_matrix
)

# ── 1. Prepare features and target ──────
X = df.drop(columns=['Target'])
y = df['Target']  # Graduate / Dropout / Enrolled

# ── 2. Train/test split ─────────────────
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# ── 3. Scale features ───────────────────
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled  = scaler.transform(X_test)

# ── 4. Train Logistic Regression ────────
lr = LogisticRegression(max_iter=1000, random_state=42)
lr.fit(X_train_scaled, y_train)

# ── 5. Predictions & Metrics ────────────
y_pred = lr.predict(X_test_scaled)

print("=" * 50)
print("BASELINE — Logistic Regression")
print("=" * 50)
print(f"\nAccuracy:  {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision: {precision_score(y_test, y_pred, average='weighted'):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred, average='weighted'):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred, average='weighted'):.4f}")
print("\nClassification Report:")
print(classification_report(y_test, y_pred))
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

BASELINE — Logistic Regression

Accuracy:  0.7684
Precision: 0.7500
Recall:    0.7684
F1 Score:  0.7531

Classification Report:
              precision    recall  f1-score   support

     Dropout       0.79      0.77      0.78       284
    Enrolled       0.52      0.33      0.41       159
    Graduate       0.80      0.93      0.86       442

    accuracy                           0.77       885
   macro avg       0.71      0.68      0.68       885
weighted avg       0.75      0.77      0.75       885

Confusion Matrix:
[[218  29  37]
 [ 43  53  63]
 [ 14  19 409]]


# Additional Models

In [10]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train, y_train)
y_pred_dt = dt.predict(X_test)
dt_accuracy = accuracy_score(y_test, y_pred_dt)
dt_precision = precision_score(y_test, y_pred_dt, average='weighted')
dt_recall = recall_score(y_test, y_pred_dt, average='weighted')
dt_f1 = f1_score(y_test, y_pred_dt, average='weighted')
print("=" * 50)
print("MODEL 2 — Decision Tree")
print("=" * 50)
print(f"\nAccuracy: {dt_accuracy:.4f}")
print(f"Precision: {dt_precision:.4f}")
print(f"Recall: {dt_recall:.4f}")
print(f"F1 Score: {dt_f1:.4f}")
print("\nClassification Report:")
print(classification_report(y_test, y_pred_dt))
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_dt))
cv_scores_dt = cross_val_score(
 DecisionTreeClassifier(random_state=42),
 X,
 y,
 cv=5,
 scoring='accuracy'
)
print("\n5-Fold Cross-Validation (Accuracy):")
print("Scores:", cv_scores_dt)
print(f"Mean CV Accuracy: {cv_scores_dt.mean():.4f}")
print(f"Std CV Accuracy: {cv_scores_dt.std():.4f}")


MODEL 2 — Decision Tree

Accuracy: 0.6972
Precision: 0.7019
Recall: 0.6972
F1 Score: 0.6994

Classification Report:
              precision    recall  f1-score   support

     Dropout       0.70      0.70      0.70       284
    Enrolled       0.40      0.43      0.41       159
    Graduate       0.81      0.79      0.80       442

    accuracy                           0.70       885
   macro avg       0.64      0.64      0.64       885
weighted avg       0.70      0.70      0.70       885

Confusion Matrix:
[[198  47  39]
 [ 47  68  44]
 [ 36  55 351]]

5-Fold Cross-Validation (Accuracy):
Scores: [0.67457627 0.68587571 0.68248588 0.66553672 0.68438914]
Mean CV Accuracy: 0.6786
Std CV Accuracy: 0.0076


In [12]:
from sklearn.svm import SVC

svm = SVC(kernel='rbf', random_state=42)
svm.fit(X_train, y_train)

y_pred_svm = svm.predict(X_test)

svm_accuracy = accuracy_score(y_test, y_pred_svm)
svm_precision = precision_score(y_test, y_pred_svm, average='weighted')
svm_recall = recall_score(y_test, y_pred_svm, average='weighted')
svm_f1 = f1_score(y_test, y_pred_svm, average='weighted')

print("=" * 50)
print("MODEL 3 — Support Vector Machine")
print("=" * 50)
print(f"\nAccuracy: {svm_accuracy:.4f}")
print(f"Precision: {svm_precision:.4f}")
print(f"Recall: {svm_recall:.4f}")
print(f"F1 Score: {svm_f1:.4f}")
print("\nClassification Report:")
print(classification_report(y_test, y_pred_svm))
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_svm))

cv_scores_svm = cross_val_score(SVC(kernel='rbf', random_state=42), X, y, cv=5, scoring='accuracy')
print("\n5-Fold Cross-Validation (Accuracy):")
print("Scores:", cv_scores_svm)
print(f"Mean CV Accuracy: {cv_scores_svm.mean():.4f}")
print(f"Std CV Accuracy: {cv_scores_svm.std():.4f}")


MODEL 3 — Support Vector Machine

Accuracy: 0.4994
Precision: 0.2494
Recall: 0.4994
F1 Score: 0.3327

Classification Report:
              precision    recall  f1-score   support

     Dropout       0.00      0.00      0.00       284
    Enrolled       0.00      0.00      0.00       159
    Graduate       0.50      1.00      0.67       442

    accuracy                           0.50       885
   macro avg       0.17      0.33      0.22       885
weighted avg       0.25      0.50      0.33       885

Confusion Matrix:
[[  0   0 284]
 [  0   0 159]
 [  0   0 442]]

5-Fold Cross-Validation (Accuracy):
Scores: [0.49830508 0.49943503 0.49943503 0.49943503 0.5       ]
Mean CV Accuracy: 0.4993
Std CV Accuracy: 0.0006


<table>
<tr>
<th>Model</th>
<th>Accuracy</th>
<th>Precision</th>
<th>Recall</th>
<th>F1-Score</th>
</tr>

<tr>
<td>Base model</td>
<td>0.768</td>
<td>0.750</td>
<td>0.768</td>
<td>0.753</td>
</tr>

<tr>
<td>Decision Tree</td>
<td>0.697</td>
<td>0.701</td>
<td>0.697</td>
<td>0.699</td>
</tr>

<tr>
<td>SVM</td>
<td>0.499</td>
<td>0.249</td>
<td>0.499</td>
<td>0.332</td>
</tr>


    
</table>

# Interpretation

## The Base Model performed best, 
achieving the highest scores across all metrics (Accuracy: 0.768, F1-Score: 0.753).

## Why it likely outperformed the others:
### vs. Decision 
Tree (0.699 F1): The base model's edge over the Decision Tree suggests the dataset likely benefits from smoother or more probabilistic decision boundaries. Decision Trees are prone to overfitting on noisy or high-dimensional data — if the dataset has overlapping classes or moderate noise, a Decision Tree's rigid axis-aligned splits will generalize worse than a model that captures softer patterns.
### vs. SVM 
(0.332 F1): The SVM's dramatic collapse in performance — despite SVMs being theoretically powerful — strongly suggests the data is not linearly separable in its current feature space, and that the kernel/hyperparameters weren't well-tuned. Alternatively, if the dataset is imbalanced, SVM's precision of 0.249 (near random) indicates it's defaulting to predicting the majority class almost exclusively.

## Key disagreement / failure observation:
The most striking finding is the SVM's precision-recall symmetry breaking: its Precision (0.249) is half its Recall (0.499), meaning it casts a very wide net — catching true positives but flooding predictions with false positives. This contrasts with the Base Model and Decision Tree, where Precision ≈ Recall, suggesting balanced decision thresholds. This asymmetry implies the SVM's decision boundary is poorly calibrated for this dataset, possibly due to feature scaling issues (SVMs are highly sensitive to unscaled features) or an untuned regularization parameter C.

# Process Log

### What We Tried
We began by establishing a baseline classifier to set a performance floor before introducing more complex models. The pipeline involved standard preprocessing — handling missing values, encoding categoricals, and an initial train/test split. We then introduced a Decision Tree to explore whether interpretable, non-linear boundaries would capture structure the base model missed. Finally, we attempted an SVM, expecting it to handle class boundaries more robustly given its margin-maximisation objective.
### What Worked
The base model came together quickly and provided a reliable reference point. Having that anchor early was useful — it meant we could immediately see whether subsequent models were actually adding value or just adding complexity. The Decision Tree was straightforward to implement and its performance was close enough to the baseline that it confirmed the dataset doesn't have dramatically complex decision boundaries.
### What Did Not Work
The SVM underperformed significantly and consumed the most debugging time. Our first attempt produced near-random precision (≈0.25), which we initially misread as a bug. After investigation, we traced the likely cause to unscaled features — SVMs are sensitive to feature magnitude, and we had not applied standardisation before fitting. We re-ran with StandardScaler but saw only marginal improvement, suggesting the kernel choice (default RBF) or the regularisation parameter C may also be mismatched to this dataset's geometry. We did not have time to run a full grid search.

### Blockers

- <b>Feature scaling oversight:</b> Applying scaling after the fact rather than inside a proper pipeline introduced a subtle data leakage risk we only noticed late. We mitigated it but it cost time.
- <b>Hyperparameter tuning scope:</b> We had planned a GridSearchCV pass over the SVM's C and gamma parameters but ran out of time. This is the most likely reason the SVM never reached its potential.
- <b>Class imbalance — unresolved:</b> We noticed the SVM's recall and precision diverged sharply, which is a typical signal of class imbalance, but we did not confirm the class distribution or apply any resampling strategy (e.g., SMOTE or class weighting). This remains an open question that could explain both the SVM's failure and the ceiling on the base model's performance.

### Honest Reflection
The base model winning is not necessarily a success — it may simply mean we did not invest enough in tuning the alternatives. A well-configured SVM on a clean, scaled dataset regularly outperforms simpler baselines. The result here likely reflects our process constraints more than it reflects the dataset's true structure.